src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_gemv.cl

```diff
@@ -0,0 +1,610 @@
+#define GEMV_OUTPUT_VEC_TYPE              MAKE_VECTOR_TYPE(OUTPUT_TYPE, 1)
+#define TO_GEMV_OUTPUT_VEC_TYPE(x)        CAT(convert_, GEMV_OUTPUT_VEC_TYPE)(x)
+#define TO_GEMV_FILTER_VEC_TYPE(x)        CAT(convert_, GEMV_FILTER_VEC_TYPE)(x)
+#define TO_GEMV_FILTER_PACKED_VEC_TYPE(x) CAT(convert_, GEMV_FILTER_PACKED_VEC_TYPE)(x)
+
+#define GEMV_INPUT_BLOCK_READ(ptr, offset)  BLOCK_READN(INPUT0_TYPE, INPUT_TILE_SIZE, ptr, offset)
+#define GEMV_FILTER_BLOCK_READ(ptr, offset) BLOCK_READN(FILTER_TYPE, 16, ptr, offset)
+
+inline int get_4bit_weight_index(int k, int n, int K, int N, int OSV) {
+    return (n / OSV) * (OSV * K / 2) + (n % OSV) + (k / 2) * OSV;
+}
+
+inline int get_4bit_weight_index_no_isv(int k, int n, int K, int N, int OSV) {
+    return (n / OSV) * (OSV * K / 2) + (k / 2) * OSV;
+}
+
+inline void thread_task_splitter(const int group_num, const int thr_num, const int thr_id, int* n_start, int* n_end) {
+    if (thr_num <= 1 || group_num == 0) {
+        *n_start = 0;
+        *n_end = group_num;
+    } else {
+        int num = (group_num + thr_num - 1) / thr_num;
+        int num_minus = num - 1;
+        int last = group_num - num_minus * thr_num;
+        *n_end = thr_id < last ? num : num_minus;
+        *n_start = thr_id <= last ? thr_id * num : last * num + (thr_id - last) * num_minus;
+    }
+    *n_end += *n_start;
+}
+
+__attribute__((intel_reqd_sub_group_size(SUBGROUP_SIZE))) KERNEL(fully_connected_gpu_gemv)(
+    OPTIONAL_SHAPE_INFO_ARG __global INPUT0_TYPE* input,
+#if DECOMPRESSION_SCALE_TERM
+    const __global DECOMPRESSION_SCALE_TYPE* scales,
+#endif
+#if DECOMPRESSION_ZP_TERM && !DECOMPRESSION_ZP_SCALAR
+    const __global DECOMPRESSION_ZP_TYPE* zps,
+#endif
+    __global OUTPUT_TYPE* output,
+    const __global FILTER_TYPE* weights
+#if BIAS_TERM
+    ,
+    const __global BIAS_TYPE* bias
+#endif
+#if HAS_FUSED_OPS_DECLS
+    ,
+    FUSED_OPS_DECLS
+#endif
+) {
+#if KERNEL_LAYOUT_OS_IS_YX_OSV16
+    // global:[N, M, 16]
+    // local: [16, 1, 16]
+    int n = get_global_id(0);              // N
+    int thr_id = get_local_id(2);          // 0~15
+    int thr_num = get_local_size(2);       // 16
+    int wi_id = get_sub_group_local_id();  // 0~15
+
+    int gk0, gk1;
+    thread_task_splitter(SCALE_GROUP_NUM, thr_num, thr_id, &gk0, &gk1);
+
+#    if DECOMPRESSION_ZP_SCALAR
+    char zp_scalar_value = (char)(DECOMPRESSION_ZP_VALUE);
+#    endif
+
+    __local float all_sum_even[16][16];  // [wi_id, thr_id]
+    __local float all_sum_odd[16][16];
+
+    // Scale layout is byfx
+    scales += n;
+#    if DECOMPRESSION_ZP_TERM && !DECOMPRESSION_ZP_SCALAR
+    zps += n;
+#    endif
+
+    float sum_all = 0;
+    for (int gk = gk0; gk < gk1; gk++) {
+        __global INPUT0_TYPE* A = input + gk * DECOMPRESSION_GROUP_SIZE;
+        const __global FILTER_TYPE* B =
+            weights + get_4bit_weight_index_no_isv(gk * DECOMPRESSION_GROUP_SIZE, n, WEIGHTS_K, WEIGHTS_N, 16);
+
+        GEMV_ACCUMULATOR_VEC_TYPE sum = 0;
```

Comment:
Did you have a chance to try it with fp16 accumulator type? As I can see, bf_tiled uses fp16 data type accumulator
MAD instruction doesn't support mixed types (like float += half * half), so fp32 prevents mad instruction usage. Or, considering the memory-bound nature, is there no difference between fp16 and fp32 accumulators from performance perspective?

Reply:
Oh, yeah, but it seems this is not real "half*half -> float" mad, but half accumulator is used to sum up 4 MAD instructions together, and then this result is written to the float accumulator